source/source.go

```diff
@@ -0,0 +1,147 @@
+	return map[string]sdk.Parameter{
+		config.KeyAccessToken: {
+			Default:     "",
+			Required:    true,
+			Description: "The private app’s access token for accessing the HubSpot API.",
+		},
+		config.KeyResource: {
+			Default:     "",
+			Required:    true,
+			Description: "The name of a HubSpot resource the connector will work with.",
+		},
+		config.KeyMaxRetries: {
+			Default:  "4",
+			Required: false,
+			Description: "The number of HubSpot API request retries " +
+				"that will be tried before giving up if a request fails.",
+		},
+		ConfigKeyPollingPeriod: {
+			Default:     "5s",
+			Required:    false,
+			Description: "The duration defines a period of polling new items if CDC is not available for a resource.",
+		},
+		ConfigKeyBufferSize: {
+			Default:  "100",
+			Required: false,
+			Description: "The buffer size for consumed items. " +
+				"It will also be used as a limit when retrieving items from the HubSpot API.",
+		},
+	}
+}
+
+// Configure parses and initializes the config.
+func (s *Source) Configure(ctx context.Context, cfg map[string]string) (err error) {
+	s.config, err = ParseConfig(cfg)
+	if err != nil {
+		return fmt.Errorf("parse source config: %w", err)
+	}
+
+	return nil
+}
+
+// Open makes sure everything is prepared to read records.
+func (s *Source) Open(ctx context.Context, sdkPosition sdk.Position) error {
+	retryableHTTPClient := retryablehttp.NewClient()
+	retryableHTTPClient.RetryMax = s.config.MaxRetries
+	retryableHTTPClient.Logger = sdk.Logger(ctx)
+
+	hubspotClient := hubspot.NewClient(s.config.AccessToken, retryableHTTPClient.StandardClient())
+
+	position, err := iterator.ParsePosition(sdkPosition)
+	if err != nil && !errors.Is(err, iterator.ErrEmptyPosition) {
+		return fmt.Errorf("parse position: %w", err)
+	}
+
+	s.iterator = iterator.NewSnapshot(ctx, iterator.SnapshotParams{
+		HubSpotClient: hubspotClient,
+		Resource:      s.config.Resource,
+		BufferSize:    s.config.BufferSize,
+		PollingPeriod: s.config.PollingPeriod,
+		Position:      position,
+	})
+
+	return nil
+}
+
+// Read fetches a new record from an iterator.
+// If there's no record the method will return the [sdk.ErrBackoffRetry].
+func (s *Source) Read(ctx context.Context) (sdk.Record, error) {
+	hasNext, err := s.iterator.HasNext(ctx)
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("has next: %w", err)
+	}
+
+	if !hasNext {
+		return sdk.Record{}, sdk.ErrBackoffRetry
+	}
+
+	record, err := s.iterator.Next(ctx)
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("get next record: %w", err)
+	}
+
+	return record, nil
+}
+
+// Ack does nothing. We don't need acks for the Snapshot or CDC iterators.
+func (s *Source) Ack(ctx context.Context, position sdk.Position) error {
+	sdk.Logger(ctx).Debug().Str("position", string(position)).Msg("got ack")
+
+	return nil
+}
+
+// Teardown does nothing.
+func (s *Source) Teardown(ctx context.Context) error {
+	sdk.Logger(ctx).Debug().Msg("got teardown")
+
```

Comment:
please add logic to close client connections, Something like
```
c.httpClient.CloseIdleConnections()
```

Reply:
If we use `github.com/hashicorp/go-retryablehttp` as a wrapper for the standard `*http.Client` it wouldn't make sense since the `retryablehttp.Transport` hasn't a `CloseIdleConnections` method.

And the `CloseIdleConnections`'s documentation says: _If the Client's Transport does not have a CloseIdleConnections method then this method does nothing._

Reply:
ok